### PR TITLE
Provides additional configuration options for logging-gelf

### LINKED ISSUE #6336: Add more config options for quarkus-logging-gelf
Currently, logging-gelf extension is not allowing to set facility and other options to log to graylog.
Please, consider add all possible settings to make it fully configurable

### PR DESCRIPTION
**Pull request by loicmathieu**
Fixes #6336

### CODE REVIEW

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java
Comment (machi1990) [inline extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java]: Maybe this should be a map of config group?
 
Something like in pseudo code

```
@ConfigItem
Map<String, AdditionalField> additionalField;
```

And the `AdditionalField` class being 

```
@ConfigGroup
AdditionalField {
 @ConfigItem
 String value;

 @ConfigItem
 String type;
}

WDYT? @loicmathieu @gsmet

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java
Comment (machi1990) [inline extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java]: Is it possible to have an enum here?

**Comment by machi1990** [comment]
Thank you @loicmathieu for this PR I have added two comments which are open for discussion.

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java
Comment (loicmathieu) [reply extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java]: I don't think an enum will be OK here.
The facility shoud be the name of the logging system, that's why the default value is `jboss-logmanager` but user can override it to whatever they want (typical use case: the name of the application)

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java
Comment (loicmathieu) [reply extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java]: I see your point, but I don't know if it will be easier to the user to configure them like this.

Current situation:
```
quarkus.log.handler.gelf.additional-fields=field1=value,field2=1
quarkus.log.handler.gelf.additional-fields-types=field1=String,field2=long
```

Your proposal:
```
quarkus.log.handler.gelf.additional-fields.field1.value=value
quarkus.log.handler.gelf.additional-fields.field1.type=String
quarkus.log.handler.gelf.additional-fields.field2.value=1
quarkus.log.handler.gelf.additional-fields.field2.type=long
```

Comment (machi1990) [reply extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java]: Thanks for the reply. 

With my proposal, the property will be in the generated doc. I think it also shows that the value and type go hand in hand (maybe that is pretty clear to some user but surely not all users). 

What happens if we provide fields value without the type or vice-versa with the current proposal?

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java
Comment (machi1990) [reply extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java]: Oh okay, thanks for the explanation.

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java
Comment (loicmathieu) [reply extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java]: - field value without type: default type will be used (discover: auto mode based on parseability of the field)
- field type without value: no idea

Comment (machi1990) [reply extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java]: Great, thanks. 


> * field value without type: default type will be used (discover: auto mode based on parseability of the field)

Great thanks, just saw it on the Javadoc documentation. 

> * field type without value: no idea

Okay, can we have a test? If it is something not desirable, I think we should throw a `ConfigurationException`.  

Let us wait and see what others think of the two proposals or if we can find another option to make these configuration knobs more appealing to the user.

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java
Comment (gsmet) [inline extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java]:
```suggestion
     * Post additional fields. E.g. `fieldName1=value1,fieldName2=value2`.
```

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java
Comment (gsmet) [inline extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java]:
```suggestion
     * E.g. `field=String,field2=double`.
```

**Review by gsmet — CHANGES_REQUESTED** [review]
I agree we should experiment with Manyanda's proposal and see how it goes. It looks cleaner and easier to understand.

Also, if the type is optional, it reads very well.

**Comment by loicmathieu** [comment]
@gsmet  @machi1990 I review the way additional fields are configured.
If it's OK for you, then I'll squashed the commits

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java
Comment (machi1990) [inline extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java]:
```suggestion
     * ```
```

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfLogHandlerRecorder.java
Comment (machi1990) [inline extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfLogHandlerRecorder.java]: We try to avoid lambdas in runtime module

**Comment by machi1990** [comment]
> @gsmet @machi1990 I review the way additional fields are configured.
> If it's OK for you, then I'll squashed the commits

I have added two comments, LGTM once they are addressed.

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java
Comment (loicmathieu) [reply extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java]: done

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfLogHandlerRecorder.java
Comment (loicmathieu) [reply extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfLogHandlerRecorder.java]: done

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/AdditionalFieldConfig.java
Comment (gsmet) [inline extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/AdditionalFieldConfig.java]:
```suggestion
     * Additional field type specification.
```

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/AdditionalFieldConfig.java
Comment (gsmet) [inline extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/AdditionalFieldConfig.java]: Maybe make it a `String` and make `discover` the `defaultValue`? This way, it's all auto-documented?

I was wondering if we could use an enum for that but I'm pretty sure the magic case handling we added will conflict with having `Long` and `long`.

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java
Comment (gsmet) [inline extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java]:
```suggestion
     * Name of the facility.
```

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java
Comment (gsmet) [inline extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java]: Markdown is not supported in Javadoc. Either use a `<pre>` or make it Asciidoc and use `@asciidoclet`.

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java
Comment (gsmet) [inline extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java]:
```suggestion
    public Map<String, AdditionalFieldConfig> additionalFields;
```
?

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java
Comment (gsmet) [inline extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java]: Please add a `@ConfigDocMapKey("field-name")` annotation.

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfLogHandlerRecorder.java
Comment (gsmet) [inline extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfLogHandlerRecorder.java]: There's no escaping needed?

**Review by gsmet — CHANGES_REQUESTED** [review]
I added a couple of suggestions.

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/AdditionalFieldConfig.java
Comment (loicmathieu) [reply extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/AdditionalFieldConfig.java]: I'm not comfortable with enum because of `Long` and `long`, or we can choose to only support one of them as both will ends up the same I think ...

If I have no value, I didn't emit any type configuration at all. If I add a default value the default type configuration will be created for each additional type. I have no string opinion on this ... as you prefere ...

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfLogHandlerRecorder.java
Comment (loicmathieu) [reply extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfLogHandlerRecorder.java]: In all examples that I saw, there was no escaping.
I don't know what will happens if a user put a special character in an additional field.
As it comes from a configuration file, there is no security risk so I propose to let it this way

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/AdditionalFieldConfig.java
Comment (gsmet) [reply extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/AdditionalFieldConfig.java]: I think I prefer having a `defaultValue`.

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java
Comment (loicmathieu) [reply extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java]: Updated to use pre

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java
Comment (loicmathieu) [reply extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java]: I was not aware of this annotation, updated to use it.

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/AdditionalFieldConfig.java
Comment (loicmathieu) [reply extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/AdditionalFieldConfig.java]: Done

File: extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java
Comment (loicmathieu) [reply extensions/logging-gelf/runtime/src/main/java/io/quarkus/logging/gelf/GelfConfig.java]: I also add a `@ConfigDocSection` it adds clarity to the configuration documentation.

**Comment by loicmathieu** [comment]
@gsmet I think I apply all your suggestions now.

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
Cool, let's wait for CI and merge.

**Pull request merged**
